dbms/src/Functions/FunctionsGrouping.h

```diff
@@ -0,0 +1,268 @@
+            groupingVec<2>(col_grouping_ids, col_res, row_num);
+            break;
+        case tipb::GroupingMode::ModeNumericSet:
+            groupingVec<3>(col_grouping_ids, col_res, row_num);
+            break;
+        default:
+            throw Exception(fmt::format("Invalid version {} in grouping function", magic_enum::enum_name(mode)));
+        };
+    }
+
+    template <int version>
+    void groupingVec(const ColumnPtr & col_grouping_ids, ColumnPtr & col_res, size_t row_num) const
+    {
+        // get arg's data container
+        const auto * grouping_col_vec = checkAndGetColumn<ColumnVector<ArgType>>(&(*col_grouping_ids));
+        if (grouping_col_vec == nullptr)
+            throw Exception("Arg's data type should be UInt64 in grouping function.");
+
+        const typename ColumnVector<ArgType>::Container & grouping_container = grouping_col_vec->getData();
+
+        // get result's data container
+        auto col_vec_res = ColumnVector<ResultType>::create();
+        typename ColumnVector<ResultType>::Container & vec_res = col_vec_res->getData();
+        vec_res.resize_fill(row_num, static_cast<ResultType>(0));
+
+        for (size_t i = 0; i < row_num; ++i)
+        {
+            if constexpr (version == 1)
+                vec_res[i] = groupingImplModeAndBit(grouping_container[i]);
+            else if constexpr (version == 2)
+                vec_res[i] = groupingImplModeNumericCmp(grouping_container[i]);
+            else if constexpr (version == 3)
+                vec_res[i] = groupingImplModeNumericSet(grouping_container[i]);
+            else
+                throw Exception("Invalid version in grouping function");
+        }
+        col_res = std::move(col_vec_res);
+    }
+
+    ResultType grouping(UInt64 grouping_id) const
+    {
+        switch (mode)
+        {
+        case tipb::GroupingMode::ModeBitAnd:
+            return groupingImplModeAndBit(grouping_id);
+        case tipb::GroupingMode::ModeNumericCmp:
+            return groupingImplModeNumericCmp(grouping_id);
+        case tipb::GroupingMode::ModeNumericSet:
+            return groupingImplModeNumericSet(grouping_id);
+        default:
+            throw Exception(fmt::format("Invalid version {} in grouping function", magic_enum::enum_name(mode)));
+        }
+    }
+
+    ResultType groupingImplModeAndBit(UInt64 grouping_id) const
+    {
+        return (grouping_id & meta_grouping_id) != 0;
+    }
+
+    ResultType groupingImplModeNumericCmp(UInt64 grouping_id) const
+    {
+        return grouping_id > meta_grouping_id;
+    }
+
+    ResultType groupingImplModeNumericSet(UInt64 grouping_id) const
+    {
+        auto iter = meta_grouping_marks.find(grouping_id);
+        return iter == meta_grouping_marks.end();
+    }
+
+private:
+    DataTypes argument_types;
+    DataTypePtr return_type;
+
+    tipb::GroupingMode mode;
+    UInt64 meta_grouping_id = 0;
+    std::set<UInt64> meta_grouping_marks = {};
```

Comment:
Why use `std::set` instead of `std::unordered_set`?

Reply:
> Why use `std::set` instead of `std::unordered_set`?

According to [here](https://stackoverflow.com/questions/1349734/why-would-anyone-use-set-instead-of-unordered-set#:~:text=For%20a%20small%20number%20of%20elements%2C%20lookups%20in%20a%20set%20might%20be%20faster%20than%20lookups%20in%20an%20unordered_set.), I think `set` is faster than `unordered_set`.

Reply:
> 

okk